libraries/type-length-value/src/unsized_pack.rs

```diff
@@ -0,0 +1,29 @@
+//! The [`UnsizedPack`] serialization trait.
+
+use solana_program::program_error::ProgramError;
+
+/// Trait that mimics a lot of the functionality of `solana_program::program_pack::Pack`
+/// but specifically works for variable-size types.
+pub trait UnsizedPack {
+    /// Writes the serialized form of the instance into the given slice
+    fn pack_into_slice(&self, dst: &mut [u8]) -> Result<(), ProgramError>;
+
+    /// Deserializes the type from the given slice
+    // The `Result` enum wrapping this requires that `Self` be sized. Maybe we
+    // can do something more clever here?
```

Comment:
Let's go with `VariableLenPack`, I think it'll make everything clearer since a `!Sized` type is not the same as a struct with a `String` or `Vec` field.